RFD 10: client and API refactoring - implemented update (#5367)
gravitational/teleport: rfd/0010-api.md

## Details

The api package will be made a Go module, with no dependency on Teleport code outside of itself. This will allow for users to use different Teleport client and server versions together, as long as they are compatibible according to the compatibility matrix. This will also allow the api to have far fewer dependencies, which both lowers the api's import footprint and possible external compatibility issues.

In addition to separating out the client from internal Teleport libraries, the client needs several quality of life improvements. These changes include better client UX via easy credential loading, automatic version checking, better (generated) docs, and more language implementations.

### Make client independent

Keep Teleport's API layer unified, avoid having internal API vs external API layer. Everything needed for a client to use the api client should live within the `api` package.

```
github.com/gravitational/teleport/api:

go.mod
go.sum
api/
  client/
    client.go
    proto/

>       authservice.proto

Reviewer: The proto file currently includes the service interface which does not belong in the api/client and will probably need to be moved elsewhere

Author: Are you referring to the `AuthServiceServer` interface? The `AuthService` defined in the proto file is just the name of the Teleport API as it was,  which connects directly to the auth service. This may be changed in the next step (API credential loading) of the refactor, in which case we may want to rename it to `TeleportAPI` more generally, but I still need to look into this.

type services.Role = types.Role
type auth.Service = proto.Service
```

**Track dependencies**

Track dependencies for the `api` package to be limited to:
  * `gravitational/trace`
  * grpc google package and its dependencies
  * protobuf generated code

External libraries for things like json and yaml encoding should be refactored to use Go's built in libraries.

**Move client implementation to `api/client`**

Move relevant gRPC related `lib/auth.Client` code to `api/client`. Move the protobuf files to `client/proto`, along with any other client code that it depends on.

Leave the `lib/auth.Client` intact for backwards compatibility, internal use, and to retain functionality of our soon-to-be deprecated http client. Necessary endpoints should be refactored into the gRPC client, rather than moving the http client to the new `api` package.

**Separate Teleport services from protobuf types**

Currently, `lib/services` holds the protobuf files and types that the client in `lib/auth` depends on. These types and proto files must be moved to `api/types` to remove the dependency on `lib/services`, as well as to draw a clear distinction between client logic and business logic.

> Many methods defined on these types are beyond the scope of an external client, such as Certificate Authority Rotation methods, and should be refactored back into services for internal use.

Reviewer: ```suggestion
Many methods defined on these types are beyond the scope of an external client, such as Certificate Authority Rotation methods, and should be moved to lib/services for internal use.
```
